File: packages/ledger/src/checksum.ts
import type {
  ChecksumCause,
  ChecksumReport,
  Confidence,
  ProposedBatch,
  Transaction,
} from "./types";

/**
 * A conferência: soma das transações extraídas contra o total declarado.
 *
 * É o diferencial do domínio — o documento carrega a própria prova. Bateu, a
 * extração está validada matematicamente, não por confiança no modelo.
 * Divergiu, temos um diagnóstico: quanto falta e onde provavelmente está.
 *
 * Sem tolerância por padrão, e de propósito: dinheiro bate ao centavo. Uma
 * tolerância "só para arredondar" é como erro de extração passa despercebido.
 */
/**
 * Quais lançamentos compõem o total declarado.
 *
 * Descoberto contra uma fatura real: o **pagamento da fatura anterior** aparece
 * na lista de lançamentos, mas NÃO entra no total desta fatura — ele quita o
 * ciclo passado. Estorno entra (é crédito contra uma compra do período);
 * encargo entra; pagamento não.
 *
 * Somar tudo indiscriminadamente produzia uma divergência do tamanho exato do
 * pagamento — R$ 3.251,62 numa fatura de R$ 4.387,92 — que parecia erro de
 * extração e era erro de modelo.
 */
export function countsTowardDeclaredTotal(transaction: Transaction): boolean {
  return transaction.kind !== "payment";
}

export function verifyChecksum(
  batch: ProposedBatch,
  toleranceCents = 0,
  rounding: RoundingPolicy = DEFAULT_ROUNDING,
): ChecksumReport {
  const counted = batch.transactions.filter(countsTowardDeclaredTotal);
  const extractedTotal = sumAmounts(counted);

  if (batch.declaredTotal === null) {
    return {
      result: "no_declared_total",
      extractedTotal,
      declaredTotal: null,
      difference: null,
      // Sem total declarado não há divergência a explicar; a conferência é
      // integralmente humana e o cartão sinaliza isso.
      suspectItems: [],
    };
  }

  const difference = extractedTotal - batch.declaredTotal;

  if (Math.abs(difference) <= toleranceCents) {
    return {
      result: "match",
      extractedTotal,
      declaredTotal: batch.declaredTotal,
      difference,
      suspectItems: [],
    };
  }

  const exactMatch = counted.some((transaction) => transaction.amount === difference);
  const likelyCause = classifyCause(difference, counted.length, exactMatch, rounding);
  const localizedIn = localize(batch, counted);

  return {
    result: "mismatch",
    likelyCause,
    ...(localizedIn === null ? {} : { localizedIn }),
    extractedTotal,
    declaredTotal: batch.declaredTotal,
    difference,
    // Arredondamento não tem culpado: listar suspeitos ali seria apontar para
    // o lugar errado com aparência de precisão.
    // Só os lançamentos que compõem o total podem explicar a divergência.
    suspectItems: likelyCause === "rounding" ? [] : rankSuspects(counted, difference),
  };
}

/**
 * Onde a diferença mora.
 *
 * Faturas declaram subtotais no resumo ("IOF de compras internacionais
 * R$ 35,17"). Comparar cada subtotal com a soma das linhas do mesmo tipo
 * transforma "a conta não bate" em "a conta não bate no IOF" — que é a
 * diferença entre um alarme e um diagnóstico.
 *
 * Verificado numa fatura real: 7 linhas de IOF somando R$ 35,16 contra
 * R$ 35,17 declarados. Sem isto, o sistema sabia que faltava 1 centavo e não
 * sabia dizer onde.
 */
function localize(
  batch: ProposedBatch,
  counted: Transaction[],
): ChecksumReport["localizedIn"] | null {
  const declared = batch.declaredSubtotals;
  if (declared === null || declared === undefined) return null;

  const sumOf = (predicate: (t: Transaction) => boolean) =>
    counted.filter(predicate).reduce((total, t) => total + t.amount, 0);

  if (declared.fees !== null && declared.fees !== undefined) {
    const extracted = sumOf((t) => t.kind === "fee");
    if (extracted !== declared.fees) {
      return { area: "fees", declared: declared.fees, extracted };
    }
  }

  if (declared.purchases !== null && declared.purchases !== undefined) {
    const extracted = sumOf((t) => t.kind === "purchase");
    if (extracted !== declared.purchases) {
      return { area: "purchases", declared: declared.purchases, extracted };
    }
  }

  return null;
}

/**
 * Poucos centavos espalhados por muitos lançamentos é assinatura de
 * arredondamento acumulado, não de item lido errado — um item errado erra por
 * ordens de grandeza maiores. O limiar cresce com a quantidade de itens porque
 * cada um pode contribuir com meio centavo.
 *
 * É heurística, e heurística com número fixo escondido no meio do arquivo é
 * exatamente o tipo de decisão que ninguém consegue revisar depois. Fica aqui,
 * nomeada e substituível: `verifyChecksum` aceita outra política, e o dia em
 * que um emissor exigir tolerância diferente não vai precisar de um `if` novo
 * dentro do cálculo.
 */
export type RoundingPolicy = {
  /** Piso, em centavos: abaixo disto a diferença é sempre arredondamento. */
  floorCents: number;
  /** Quanto cada item pode contribuir, em centavos. */
  perItemCents: number;
};

export const DEFAULT_ROUNDING: RoundingPolicy = { floorCents: 2, perItemCents: 0.5 };

function classifyCause(
  difference: number,
  itemCount: number,
  hasExactMatch: boolean,
  rounding: RoundingPolicy,
): ChecksumCause {
  if (hasExactMatch) return "item";

  const tolerance = Math.max(rounding.floorCents, Math.ceil(itemCount * rounding.perItemCents));
  if (Math.abs(difference) <= tolerance) return "rounding";

  return "unknown";
}

/** Soma em inteiros. Nenhum ponto flutuante toca em dinheiro. */
export function sumAmounts(transactions: Transaction[]): number {
  return transactions.reduce((total, transaction) => total + transaction.amount, 0);
}

const CONFIDENCE_RANK: Record<Confidence, number> = { baixa: 0, media: 1, alta: 2 };

/**
 * Ordena os candidatos prováveis da divergência.
 *
 * A heurística tem uma ordem deliberada:
 *  1. Um item cujo valor é exatamente a diferença é quase certamente a causa
 *     (item lido a mais, ou a menos). Vem primeiro, qualquer que seja a
 *     confiança.
 *  2. Depois, menor confiança primeiro — é onde o extrator já se declarou
 *     inseguro.
 *  3. Empate resolve por valor absoluto: erro grande importa mais.
 */
function rankSuspects(
  transactions: Transaction[],
  difference: number,
): ChecksumReport["suspectItems"] {
  const candidates = transactions.map((transaction) => {
    const exact = transaction.amount === difference;
    return {
      transaction,
      exact,
      reason: exact
        ? "o valor deste item é exatamente a diferença — provável leitura duplicada ou faltante"
        : `confiança ${transaction.extractionConfidence} na extração`,
    };
  });

  const sorted = candidates.sort((a, b) => {
    if (a.exact !== b.exact) return a.exact ? -1 : 1;
    const byConfidence =
      CONFIDENCE_RANK[a.transaction.extractionConfidence] -
      CONFIDENCE_RANK[b.transaction.extractionConfidence];
    if (byConfidence !== 0) return byConfidence;
    return Math.abs(b.transaction.amount) - Math.abs(a.transaction.amount);
  });

  const flagged = sorted.filter(
    (candidate) => candidate.exact || candidate.transaction.extractionConfidence !== "alta",
  );

  // Quando o extrator se declarou seguro de tudo e ainda assim não fecha, o
  // erro está em algum lugar — devolver lista vazia diante de uma divergência
  // grande deixaria a pessoa sem por onde começar. Nesse caso mostramos os
  // maiores valores, que são onde um erro pesa mais.
  const chosen =
    flagged.length > 0
      ? flagged
      : sorted.map((candidate) => ({
          ...candidate,
          reason: "nada se destacou; este é um dos maiores valores do lote",
        }));

  return chosen
    .slice(0, 10)
    .map((candidate) => ({
      transactionId: candidate.transaction.id,
      reason: candidate.reason,
      confidence: candidate.transaction.extractionConfidence,
      amount: candidate.transaction.amount,
      page: candidate.transaction.page,
    }));
}

const BRL_FORMATTER = new Intl.NumberFormat("pt-BR", {
  style: "currency",
  currency: "BRL",
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
});

/** Formata centavos em `R$ #.###,##` para exibição. Nunca usado em cálculo. */
export function formatCents(cents: number): string {
  if (!Number.isInteger(cents)) {
    throw new TypeError("formatCents exige um valor inteiro em centavos.");
  }
  // `Intl` usa NBSP entre a moeda e o número. Visualmente é igual, mas
  // normalizar para espaço deixa cópia, snapshot e testes no mesmo padrão.
  return BRL_FORMATTER.format(cents / 100).replace(/\u00a0/g, " ");
}

File: services/agent/clara/ledger/checksum.py
"""A conferência: soma extraída contra o total declarado. Porta de `checksum.ts`.

É o diferencial do domínio — o documento carrega a própria prova. Bateu, a
extração está validada matematicamente, não por confiança no modelo. Divergiu,
há um diagnóstico: quanto falta e onde provavelmente está.

Sem tolerância por padrão, de propósito: dinheiro bate ao centavo. Uma
tolerância "só para arredondar" é como erro de extração passa despercebido.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from clara.ledger.types import ChecksumCause, ChecksumResult, Confidence, TransactionLike

_NON_SPEND_KINDS = {"payment", "card_payment", "transfer", "income"}


def counts_toward_declared_total(transaction: TransactionLike) -> bool:
    """Pagamento da fatura anterior aparece na lista de lançamentos mas NÃO
    entra no total desta fatura — ele quita o ciclo passado. Estorno entra
    (crédito contra compra do período); encargo entra; pagamento não."""
    return transaction["kind"] not in _NON_SPEND_KINDS


@dataclass(frozen=True)
class RoundingPolicy:
    """Heurística nomeada e substituível — não um número fixo escondido no cálculo."""

    floor_cents: int = 2
    per_item_cents: float = 0.5


DEFAULT_ROUNDING = RoundingPolicy()


def rounding_tolerance(item_count: int, rounding: RoundingPolicy = DEFAULT_ROUNDING) -> int:
    """A folga, em centavos, para um documento com este número de itens."""
    import math

    return max(rounding.floor_cents, math.ceil(item_count * rounding.per_item_cents))


@dataclass
class SuspectItem:
    transaction_id: str
    reason: str
    confidence: Confidence
    amount: int
    page: int | None


@dataclass
class LocalizedIn:
    area: Literal["fees", "purchases"]
    declared: int
    extracted: int


@dataclass
class ChecksumReport:
    result: ChecksumResult
    extracted_total: int
    declared_total: int | None
    difference: int | None
    suspect_items: list[SuspectItem] = field(default_factory=list)
    likely_cause: ChecksumCause | None = None
    tolerance: dict[str, int] | None = None
    localized_in: LocalizedIn | None = None


_CONFIDENCE_RANK: dict[Confidence, int] = {"baixa": 0, "media": 1, "alta": 2}


def _classify_cause(
    difference: int, item_count: int, has_exact_match: bool, rounding: RoundingPolicy
) -> ChecksumCause:
    if has_exact_match:
        return "item"
    if abs(difference) <= rounding_tolerance(item_count, rounding):
        return "rounding"
    return "unknown"


def _rank_suspects(transactions: list[TransactionLike], difference: int) -> list[SuspectItem]:
    """1) valor exatamente a diferença primeiro; 2) menor confiança primeiro;
    3) empate por valor absoluto — erro grande importa mais."""

    def sort_key(t: TransactionLike) -> tuple[int, int, int]:
        exact = t["amount"] == difference
        return (0 if exact else 1, _CONFIDENCE_RANK[t["extraction_confidence"]], -abs(t["amount"]))

    ordered = sorted(transactions, key=sort_key)

    def reason_for(t: TransactionLike) -> str:
        if t["amount"] == difference:
            return (
                "o valor deste item é exatamente a diferença — "
                "provável leitura duplicada ou faltante"
            )
        return f"confiança {t['extraction_confidence']} na extração"

    flagged = [
        t for t in ordered
        if t["amount"] == difference or t["extraction_confidence"] != "alta"
    ]

    if flagged:
        chosen = [(t, reason_for(t)) for t in flagged]
    else:
        # Extrator seguro de tudo e ainda assim não fecha: mostra os maiores
        # valores, que são onde um erro pesa mais, em vez de lista vazia.
        chosen = [(t, "nada se destacou; este é um dos maiores valores do lote") for t in ordered]

    return [
        SuspectItem(
            transaction_id=t["id"],
            reason=reason,
            confidence=t["extraction_confidence"],
            amount=t["amount"],
            page=t.get("page"),
        )
        for t, reason in chosen[:10]
    ]


def _localize(
    declared_subtotals: dict[str, int | None] | None, counted: list[TransactionLike]
) -> LocalizedIn | None:
    """Faturas declaram subtotais no resumo. Comparar cada um com a soma das
    linhas do mesmo tipo transforma "a conta não bate" em "a conta não bate
    no IOF" — a diferença entre um alarme e um diagnóstico."""
    if not declared_subtotals:
        return None

    def sum_of(kind: str) -> int:
        return sum(t["amount"] for t in counted if t["kind"] == kind)

    fees = declared_subtotals.get("fees")
    if fees is not None:
        extracted = sum_of("fee")
        if extracted != fees:
            return LocalizedIn(area="fees", declared=fees, extracted=extracted)

    purchases = declared_subtotals.get("purchases")
    if purchases is not None:
        extracted = sum_of("purchase")
        if extracted != purchases:
            return LocalizedIn(area="purchases", declared=purchases, extracted=extracted)

    return None


def verify_checksum(
    *,
    transactions: list[TransactionLike],
    declared_total: int | None,
    declared_subtotals: dict[str, int | None] | None = None,
    tolerance_cents: int = 0,
    rounding: RoundingPolicy = DEFAULT_ROUNDING,
) -> ChecksumReport:
    counted = [t for t in transactions if counts_toward_declared_total(t)]
    extracted_total = sum(t["amount"] for t in counted)

    if declared_total is None:
        return ChecksumReport(
            result="no_declared_total",
            extracted_total=extracted_total,
            declared_total=None,
            difference=None,
        )

    difference = extracted_total - declared_total

    if abs(difference) <= tolerance_cents:
        return ChecksumReport(
            result="match",
            extracted_total=extracted_total,
            declared_total=declared_total,
            difference=difference,
        )

    exact_match = any(t["amount"] == difference for t in counted)
    likely_cause = _classify_cause(difference, len(counted), exact_match, rounding)
    localized_in = _localize(declared_subtotals, counted)

    tolerance = None
    if likely_cause == "rounding":
        tolerance = {
            "cents": rounding_tolerance(len(counted), rounding),
            "item_count": len(counted),
        }

    suspect_items = [] if likely_cause == "rounding" else _rank_suspects(counted, difference)

    return ChecksumReport(
        result="mismatch",
        extracted_total=extracted_total,
        declared_total=declared_total,
        difference=difference,
        likely_cause=likely_cause,
        tolerance=tolerance,
        localized_in=localized_in,
        suspect_items=suspect_items,
    )
